src/app-layer-http2.c

```diff
@@ -0,0 +1,71 @@
+/* Copyright (C) 2020 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+/**
+ * \file
+ *
+ * \author Philippe Antoine <p.antoine@catenacyber.fr>
+ *
+ * Parser for HTTP2, RFC 7540
+ */
+
+#include "suricata-common.h"
+#include "stream.h"
+#include "conf.h"
+
+#include "util-unittest.h"
+
+#include "app-layer-detect-proto.h"
+#include "app-layer-parser.h"
+
+#include "app-layer-http2.h"
+#include "rust.h"
+
+static int HTTP2RegisterPatternsForProtocolDetection(void)
+{
+    //TODO is this too restrictive and can be evaded ?
+    /* TODO investigate why this makes AppLayerTest09 fail/leak
+     * and not leak if we only give a 5-bytes pattern
+    if (AppLayerProtoDetectPMRegisterPatternCI(IPPROTO_TCP, ALPROTO_HTTP2,
```

Comment:
Any thoughts on what is going on here ?

Reply:
(AppLayerTest10 and AppLayerTest11 fail too)
Commenting out `    FAIL_IF(!FLOW_IS_PM_DONE(&f, STREAM_TOSERVER));` makes them succeed again

Reply:
By having a long pattern you force the protocol detection engine to post-pone its judgement about incoming data until its been able this pattern as well.

Reply:
Indeed.
But is it good or bad ?

Should not the test end protocol detection by closing the stream anyways ?